src/devicecode.rs

```diff
@@ -0,0 +1,278 @@
+use std::collections::HashMap;
+use std::fmt::Error as FormatterError;
+use std::fmt::{Debug, Display, Formatter};
+use std::time::Duration;
+
+use serde::{Deserialize, Serialize};
+
+use super::{
+    DeviceCode, EndUserVerificationUrl, ErrorResponseType, StandardErrorResponse, UserCode,
+};
+
+/// The minimum amount of time in seconds that the client SHOULD wait
+/// between polling requests to the token endpoint.  If no value is
+/// provided, clients MUST use 5 as the default.
+fn default_devicecode_interval() -> u64 {
+    5
+}
+
+///
+/// Standard OAuth2 device authorization response.
+///
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct DeviceAuthorizationDetails {
+    /// The device verification code.
+    device_code: DeviceCode,
+
+    /// The end-user verification code.
+    user_code: UserCode,
+
+    /// The end-user verification URI on the authorization The URI should be
+    /// short and easy to remember as end users will be asked to manually type
+    /// it into their user agent.
+    #[serde(alias = "verification_url")]
+    verification_uri: EndUserVerificationUrl,
+
+    /// A verification URI that includes the "user_code" (or other information
+    /// with the same function as the "user_code"), which is designed for
+    /// non-textual transmission.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    verification_uri_complete: Option<EndUserVerificationUrl>,
+
+    /// The lifetime in seconds of the "device_code" and "user_code".
+    expires_in: u64,
+
+    /// The minimum amount of time in seconds that the client SHOULD wait
+    /// between polling requests to the token endpoint.  If no value is
+    /// provided, clients MUST use 5 as the default.
+    #[serde(default = "default_devicecode_interval")]
+    interval: u64,
+
+    #[serde(flatten)]
+    extra_fields: HashMap<String, serde_json::Value>,
```

Comment:
I tried to do this and the code gets a little worse, in my opinion.

Here's what I've tried:

- Create a trait exposing the values on a Device Authorization Response called `DeviceAuthorizationResponseCommon`
- Create a struct `StandardDeviceAuthorizationResponse<EF> where EF: ExtraDeviceAuthorizationFields`
- Impl `DeviceAuthorizationResponseCommon` for `StandardDeviceAuthorizationResponse<EF>`
- Bound all functions which previously cared about `DeviceAuthorizationResponse` to be bound on `DR: DeviceAuthorizationResponseCommon`
- Changed the return from `request` and `request_async` from `DeviceAuthorizationResponse` to `DO: DeserializedOwned` so the return type is not explicitly typed.

This means:
- I have to annotate the type of `details` in the client code:
```
let details: DeviceAuthorizationResponse = device_client
        .exchange_device_code()
        .add_scope(Scope::new("profile".to_string()))
        .request(http_client)
        .expect("Failed to request codes from device auth endpoint");
```
but because we're also running methods on Details, I have to import the `DeviceAuthorizationResponseCommon` trait as well.

Perhaps there's some other way around it?

Reply:
ah, the `TokenResponse` case is slightly more complicated than we probably need here because it uses a separate trait. this was originally done to allow clients to support broken servers that were returning different kinds of responses that were incompatible with `StandardTokenResponse`'s strict typing. however, that introduced quite a bit of complexity, so I've stayed away from the trait approach for most other things (e.g., most of the `openidconnect` crate). for this case, I think we can get away with just having a struct be returned (albeit with one generic type parameter). apologies for not being clear about what I had in mind before.

I think something like this should work:
```rust
pub trait ExtraDeviceAuthorizationFields: DeserializeOwned + Debug + Serialize {}

#[derive(Clone, Debug, Deserialize, Serialize)]
struct EmptyExtraDeviceAuthorizationFields;
impl ExtraDeviceAuthorizationFields for EmptyExtraDeviceAuthorizationFields {}

#[derive(Clone, Debug, Deserialize, Serialize)]
// Not 100% sure if this default for `EF` works:
struct DeviceAuthorizationResponse<EF = EmptyExtraDeviceAuthorizationFields>
where EF: ExtraDeviceAuthorizationFields {
  device_code: DeviceCode,
  ...
  #[serde(bound = "EF: ExtraDeviceAuthorizationFields", flatten)]
  extra_fields: EF,
}
```

for reference, this is similar to how `UserInfoRequest`/`UserInfoClaims` works in the `openidconnect` crate: https://docs.rs/openidconnect/1.0.1/openidconnect/struct.UserInfoRequest.html#method.request. in this case, `EF` will become a type parameter of `DeviceAuthorizationRequest::request[_async]`

> I have to annotate the type of details in the client code:

I haven't tested, but hopefully setting the default generic type `EF = EmptyExtraDeviceAuthorizationFields` will avoid needing to specifically annotate the type at each call site. I still think it's worth this approach though, even if a type annotation is needed (an alternative caller syntax I prefer is `.request::<EmptyExtraDeviceAuthorizationFields>(http_client)`).

Reply:
Second time around this went a bit smoother. I've implemented as:
```rust
pub type StandardDeviceAuthorizationResponse = DeviceAuthorizationResponse<EmptyExtraDeviceAuthorizationFields>;
```
which just requires
```rust
//! let details: StandardDeviceAuthorizationResponse = client
//!     .exchange_device_code()
//!     .add_scope(Scope::new("read".to_string()))
//!     .request(http_client)?;
```
and to prove you can store the fields if necessary (some providers provide a `message` field to print out to screen), I added this to the Google example:
```
struct StoringFields(HashMap<String, serde_json::Value>);

impl ExtraDeviceAuthorizationFields for StoringFields {}
type StoringDeviceAuthorizationResponse = DeviceAuthorizationResponse<StoringFields>;
```
I think this is the best we're going to get right now.